### ISSUE TITLE
cfg.DIST_ENABLE false fail

### ISSUE THREAD

**Issue by bhack** [yoxu515/aot-benchmark#36]
With `cfg.DIST_ENABLE` false all the distributed specific parts are not wrapped by a condition to check if distributed was enabled or not.

**Comment by z-x-yang** [comment]
`self.DIST_ENABLE` is necessary for multi-GPU training.

**Comment by bhack** [comment]
I meant that I am trying to test single GPU.
In many places `cfg.DIST_ENABLE` is checked to safely go through the non distributed code path. 

But in many other places not:

https://github.com/yoxu515/aot-benchmark/blob/1c3a5ec51d81f3e17ff9092aa1e830206d766132/networks/managers/trainer.py#L342-L352

**Comment by bhack** [comment]
E.g. here instead the code is checking  `cfg.DIST_ENABLE`

https://github.com/yoxu515/aot-benchmark/blob/1c3a5ec51d81f3e17ff9092aa1e830206d766132/networks/managers/trainer.py#L59-L83

**Comment by z-x-yang** [comment]
The distributed sampler is useless and meaningless for evaluation, where GPUs are asynchronous instead of synchronous. The video lengths are always different for different GPUs.

**Comment by bhack** [comment]
Is this not in the trainer?
I am trying to train a single GPU job with `cfg.DIST_ENABLE=False`

**Comment by z-x-yang** [comment]
This config is designed for training.

**Comment by bhack** [comment]
Yes it is what I meant. Are we not going to have issue if we don't conditional wrap `torch.nn.parallel.DistributedDataParallel` in the trainer?

**Comment by z-x-yang** [comment]
I suppose it should be ok.

**Comment by bhack** [comment]
Isn't that one going to require `init_process_group`? But it is conditional wrapped in the trainer.

https://github.com/yoxu515/aot-benchmark/blob/1c3a5ec51d81f3e17ff9092aa1e830206d766132/networks/managers/trainer.py#L59-L64

**Comment by z-x-yang** [comment]
Sorry, I don't understand what you mean.

**Comment by bhack** [comment]
E.g. with `self.DIST_ENABLE = False` in `configs/default.py` we are going to fail directly at:
https://github.com/yoxu515/aot-benchmark/blob/1c3a5ec51d81f3e17ff9092aa1e830206d766132/networks/managers/trainer.py#L342

```python
RuntimeError: Default process group has not been initialized, please make sure to call init_process_group.
```

Cause:
https://github.com/yoxu515/aot-benchmark/issues/36#issuecomment-1458140508

As we don't have completely covered `cfg.DIST_ENABLE` safeguards and the relatrive alternative code path.

**Comment by bhack** [comment]
> Sorry, I don't understand what you mean.

Is it clear now?

**Comment by z-x-yang** [comment]
I have updated the `trainer.py`, and it should be ok to set `self.DIST_ENABLE = False` and train with a single GPU.

**Comment by bhack** [comment]
Thanks, I've checked your changes and they were the same I've done locally on my side in these days.

What do you think now about this (https://github.com/pytorch/pytorch/issues/37444)? 
https://github.com/yoxu515/aot-benchmark/blob/d5bd73fd09a82ed6c9dd8d516ab23b6a1e3f8045/tools/train.py#L78-L79

**Comment by bhack** [comment]
Other then my previous comment I think that we still have now an issue with the `keys` with `DIST_ENABLE=False`:

https://github.com/yoxu515/aot-benchmark/blob/d5bd73fd09a82ed6c9dd8d516ab23b6a1e3f8045/networks/managers/trainer.py#L677-L682

```python
    for key in boards['image'].keys():
AttributeError: 'list' object has no attribute 'keys'
```

**Comment by z-x-yang** [comment]
It should be ok now.

**Comment by bhack** [comment]
It seems that we have two issues:
- The first is that the trainer it seems to be "randomly" deadlocked on different runs same code
- images in `img_logs` seems to be not correct anymore as they are only the binary masks instead they was "composited" if I remember correctly.

**Comment by bhack** [comment]
For the first issue this is the stacktrace of one of the deadlock and it could be related to https://github.com/yoxu515/aot-benchmark/issues/36#issuecomment-1460151286:

```python
  File "/opt/conda/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 240, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/opt/conda/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 198, in start_processes
    while not context.join():
  File "/opt/conda/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 109, in join
    ready = multiprocessing.connection.wait(
  File "/opt/conda/lib/python3.10/multiprocessing/connection.py", line 936, in wait
    ready = selector.select(timeout)
  File "/opt/conda/lib/python3.10/selectors.py", line 416, in select
    fd_event_list = self._selector.poll(timeout)
````

**Comment by z-x-yang** [comment]
I guess the first problem is due to `torch.spawn`, and I have modified the code related to [it](https://github.com/yoxu515/aot-benchmark/blob/f9a62b60e12218019c0617efa1315dcc73147cdb/tools/train.py#L79). Please take a try, and hope this will work for you.

As to the second issue, `img_logs` should be images where target objects are marked with colorful masks, referring to [here](https://github.com/yoxu515/aot-benchmark/blob/f9a62b60e12218019c0617efa1315dcc73147cdb/networks/managers/trainer.py#L624).

**Comment by bhack** [comment]
> I guess the first problem is due to torch.spawn, and I have modified the code related to [it](https://github.com/yoxu515/aot-benchmark/blob/f9a62b60e12218019c0617efa1315dcc73147cdb/tools/train.py#L79). Please take a try, and hope this will work for you.

Yes, it seems ok now.

> As to the second issue, img_logs should be images w[here](https://github.com/yoxu515/aot-benchmark/blob/f9a62b60e12218019c0617efa1315dcc73147cdb/networks/managers/trainer.py#L624) target objects are marked with colorful masks, referring to here.

Yes sorry this was a false positive related to the image range in the compositing phase, thanks for the double check.

**Comment by bhack** [comment]
The same need to be fixed in the `PAOT` branch.
